nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/QueryDatabaseTable.java

```diff
@@ -0,0 +1,608 @@
+                    try {
+                        logger.debug("Executing query {}", new Object[]{selectQuery});
+                        final ResultSet resultSet = st.executeQuery(selectQuery);
+                        // Max values will be updated in the state property map by the callback
+                        final MaxValueResultSetRowCollector maxValCollector = new MaxValueResultSetRowCollector(statePropertyMap, preProcessStrategy);
+                        nrOfRows.set(JdbcCommon.convertToAvroStream(resultSet, out, tableName, maxValCollector));
+
+                    } catch (final SQLException e) {
+                        throw new ProcessException("Error during database query or conversion of records to Avro", e);
+                    }
+                }
+            });
+
+            if (nrOfRows.get() > 0) {
+                // set attribute how many rows were selected
+                fileToProcess = session.putAttribute(fileToProcess, RESULT_ROW_COUNT, nrOfRows.get().toString());
+
+                logger.info("{} contains {} Avro records; transferring to 'success'",
+                        new Object[]{fileToProcess, nrOfRows.get()});
+                session.getProvenanceReporter().modifyContent(fileToProcess, "Retrieved " + nrOfRows.get() + " rows",
+                        stopWatch.getElapsed(TimeUnit.MILLISECONDS));
+                session.transfer(fileToProcess, REL_SUCCESS);
+            } else {
+                // If there were no rows returned, don't send the flowfile
+                session.remove(fileToProcess);
+                context.yield();
+            }
+
+        } catch (final Throwable /*ProcessException | SQLException*/ e) {
+            if (fileToProcess == null) {
+                // This can happen if any exceptions occur while setting up the connection, statement, etc.
+                logger.error("Unable to execute SQL select query {} due to {}. No FlowFile to route to failure",
+                        new Object[]{selectQuery, e});
+                context.yield();
+            } else {
+
+                logger.error("Unable to execute SQL select query {} due to {}; routing to failure",
+                        new Object[]{selectQuery, e});
+                session.transfer(fileToProcess, REL_FAILURE);
+                context.yield();
+            }
+        } finally {
+            try {
+                session.commit();
+                // Update the state
+                stateManager.setState(statePropertyMap, Scope.CLUSTER);
+            } catch (IOException ioe) {
+                getLogger().error("{} failed to process due to {}; rolling back session", new Object[]{this, ioe});
+                session.rollback(true);
+            } catch (Throwable t) {
+                getLogger().error("{} failed to process due to {}; rolling back session", new Object[]{this, t});
+                session.rollback(true);
```

Comment:
In this case, the Exception was thrown either by setState or session.commit. If session.commit, then the session will have already rolled itself back before throwing an Exception. If setState, then the session commit has already completed, so the rollback doesn't really do anything for us. Would recommend just logging the fact that the state could not be saved and why, and then continue on.